ports/stm32/machine_i2s.c

```diff
@@ -0,0 +1,1109 @@
+    if (ring_buffer_len > 0) {
+        uint8_t *buffer = m_new(uint8_t, ring_buffer_len);
+        self->ring_buffer_storage = buffer;
+        ringbuf_init(&self->ring_buffer, buffer, ring_buffer_len);
+    } else {
+        mp_raise_ValueError(MP_ERROR_TEXT("invalid buf"));
+    }
+
+    self->sck = MP_OBJ_TO_PTR(args[ARG_sck].u_obj);
+    self->ws = MP_OBJ_TO_PTR(args[ARG_ws].u_obj);
+    self->sd = MP_OBJ_TO_PTR(args[ARG_sd].u_obj);
+    self->mode = i2s_mode;
+    self->bits = i2s_bits;
+    self->format = i2s_format;
+    self->rate = args[ARG_rate].u_int;
+    self->ibuf = ring_buffer_len;
+    self->callback_for_non_blocking = MP_OBJ_NULL;
+    self->non_blocking_descriptor.copy_in_progress = false;
+    self->io_mode = BLOCKING;
+
+    I2S_InitTypeDef *init = &self->hi2s.Init;
+    init->Mode = i2s_mode;
+    init->Standard = I2S_STANDARD_PHILIPS;
+    init->DataFormat = get_dma_bits(self->mode, self->bits);
+    init->MCLKOutput = I2S_MCLKOUTPUT_DISABLE;
+    init->AudioFreq = args[ARG_rate].u_int;
+    init->CPOL = I2S_CPOL_LOW;
+    init->ClockSource = I2S_CLOCK_PLL;
+
+    // init the I2S bus
+    if (!i2s_init(self)) {
+        mp_raise_msg_varg(&mp_type_OSError, MP_ERROR_TEXT("I2S init failed"));
+    }
+
+    // start DMA.  DMA is configured to run continuously, using a circular buffer configuration
+    uint32_t number_of_samples = 0;
+    if (init->DataFormat == I2S_DATAFORMAT_16B) {
+        number_of_samples = SIZEOF_DMA_BUFFER_IN_BYTES / sizeof(uint16_t);
+    } else {  // 32 bits
+        number_of_samples = SIZEOF_DMA_BUFFER_IN_BYTES / sizeof(uint32_t);
+    }
+
+    HAL_StatusTypeDef status;
+    if (self->mode == I2S_MODE_MASTER_TX) {
+        status = HAL_I2S_Transmit_DMA(&self->hi2s, (void *)self->dma_buffer_dcache_aligned, number_of_samples);
+    } else {  // RX
+        status = HAL_I2S_Receive_DMA(&self->hi2s, (void *)self->dma_buffer_dcache_aligned, number_of_samples);
+    }
+
+    if (status != HAL_OK) {
+        mp_raise_msg_varg(&mp_type_OSError, MP_ERROR_TEXT("DMA init failed"));
+    }
+}
+
+STATIC void machine_i2s_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    machine_i2s_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "I2S(id=%u,\n"
+        "sck="MP_HAL_PIN_FMT ",\n"
+        "ws="MP_HAL_PIN_FMT ",\n"
+        "sd="MP_HAL_PIN_FMT ",\n"
+        "mode=%u,\n"
+        "bits=%u, format=%u,\n"
+        "rate=%d, ibuf=%d)",
+        self->i2s_id,
+        mp_hal_pin_name(self->sck),
+        mp_hal_pin_name(self->ws),
+        mp_hal_pin_name(self->sd),
+        self->mode,
+        self->bits, self->format,
+        self->rate, self->ibuf
+        );
+}
+
+STATIC mp_obj_t machine_i2s_make_new(const mp_obj_type_t *type, size_t n_pos_args, size_t n_kw_args, const mp_obj_t *args) {
+    mp_arg_check_num(n_pos_args, n_kw_args, 1, MP_OBJ_FUN_ARGS_MAX, true);
+    uint8_t i2s_id = mp_obj_get_int(args[0]);
+    machine_i2s_obj_t *self = m_new_obj(machine_i2s_obj_t);
```

Comment:
Following how other peripherals work, if `machine_i2s_obj[id]` already exists then that instance should be reused instead of allocating a new one.

Reply:
fixed in https://github.com/miketeachman/micropython/commit/0f7b59c92e8173d86f596097f9f8fc5ef3e3cfd5